patmat bugfixes and minor clean ups

adriaanm opened the pull request:
review by @retronym, puh-lease

Commit:
exhaust unit: consider Unit as sealed

one of the most boring sealed types imaginable, but a sealed type all the same

Commit:
minor cleanup in patmat and typers

patmat:
- remove Var's unused 'checked' member; subsumed by 'domain'
- updated docs: no longer using isSuccess in __match

typers: clean up indentation, swap ifs to failure-first

Commit:
better fix for SI-5189 pt1

the fix in 0cffdf38d9 was brain dead (only looking at the first argument of the ctor, rather than the first argument list)
the reasoning was that we should not look at the MethodType of the constructor,
but at the argument types directly (since going through the method type would flip variance,
though the ctor arguments are conceptually in covariant positions here)

Commit:
SI-2442 fixed by virtpatmat -- test files only

Commit:
SI-5914 handle null in classtag extractor

Commit:
SI-5899 exhaustiveness for non-class types

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/349/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/349/

retronym commented:
That's all I can spot, otherwise lgtm.

Commit:
make tests independent of compiler api

TODO: t5899 should also be refactored, but I couldn't figure out how
I tried the obvious Cake Light pattern with abstract types etc, but that didn't trigger it
there must be something with indirection through paths as well

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/356/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/356/
